src/Features/Core/Portable/ExternalAccess/Watch/Api/WatchHotReloadService.cs

```diff
@@ -0,0 +1,91 @@
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.EditAndContinue;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Debugger.Contracts.EditAndContinue;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Watch.Api
+{
+    internal sealed class WatchHotReloadService
+    {
+        private sealed class DebuggerService : IManagedEditAndContinueDebuggerService
+        {
+            public static readonly DebuggerService Instance = new();
+
+            public Task<ImmutableArray<ManagedActiveStatementDebugInfo>> GetActiveStatementsAsync(CancellationToken cancellationToken)
+                => Task.FromResult(ImmutableArray<ManagedActiveStatementDebugInfo>.Empty);
+
+            public Task<ManagedEditAndContinueAvailability> GetAvailabilityAsync(Guid module, CancellationToken cancellationToken)
+                => Task.FromResult(new ManagedEditAndContinueAvailability(ManagedEditAndContinueAvailabilityStatus.Available));
+
+            public Task PrepareModuleForUpdateAsync(Guid module, CancellationToken cancellationToken)
+                => Task.CompletedTask;
+        }
+
+        public readonly struct Update
+        {
+            public readonly Guid ModuleId;
+            public readonly ImmutableArray<byte> ILDelta;
+            public readonly ImmutableArray<byte> MetadataDelta;
+            public readonly ImmutableArray<byte> PdbDelta;
+
+            public Update(Guid moduleId, ImmutableArray<byte> ilDelta, ImmutableArray<byte> metadataDelta, ImmutableArray<byte> pdbDelta)
+            {
+                ModuleId = moduleId;
+                ILDelta = ilDelta;
+                MetadataDelta = metadataDelta;
+                PdbDelta = pdbDelta;
+            }
+        }
+
+        private static readonly SolutionActiveStatementSpanProvider s_solutionActiveStatementSpanProvider =
+            (_, _) => ValueTaskFactory.FromResult(ImmutableArray<TextSpan>.Empty);
+
+        private readonly IEditAndContinueWorkspaceService _encService;
+
+        public WatchHotReloadService(HostWorkspaceServices services)
+            => _encService = services.GetRequiredService<IEditAndContinueWorkspaceService>();
+
+        /// <summary>
+        /// Starts the watcher.
+        /// </summary>
+        /// <param name="solution">Solution that represents sources that match the built binaries on disk.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        /// <returns></returns>
+        public async Task StartSessionAsync(Solution solution, CancellationToken cancellationToken)
+            => await _encService.StartDebuggingSessionAsync(solution, captureMatchingDocuments: true, cancellationToken).ConfigureAwait(false);
+
+        public async Task<(ImmutableArray<Update> updates, ImmutableArray<DiagnosticData> diagnostics)> EmitSolutionUpdateAsync(Solution solution, CancellationToken cancellationToken)
```

Comment:
I think DiagnosticData is internal to the workspaces assembly. We would have to shim it too for watch to be able to access it.

Reply:
You're right. Will fix.

Reply:
What information for each rude edit do you need? Id, Message, Severity, Line & File info?

Reply:
Actually, never mind. Will return `Diagnostic`s.